### PR TITLE
Bugfix for DateTime filters in reports

### PR DESCRIPTION
**Pull request by virlatinus**
| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Related user documentation PR URL | N/A
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | N/A
| BC breaks? | N/A
| Deprecations? | N/A

#### Description:
There is an issue with DateTime filters in reports, because in the backend the filters are applied using like, so the generated query is like this one: 

```
SELECT 
    l.id AS id,
    l.email AS email,
    l.firstname AS firstname,
    l.lastname AS lastname
FROM
    leads l
WHERE
    (l.date_added BETWEEN '2017-07-13 00:00:00' AND '2017-08-12 01:58:27')
        AND (l.date_identified LIKE '2017-07-13 19:50:07')
LIMIT 30 OFFSET 0
```
This PR fixes this by fetching the report filters and replacing LIKE with the correct operator.

#### Steps to reproduce the bug:
1. Create a contact report with a dynamic filter of "date identified >= 1/1/2017".  
2. Close and save report
3. Enter a new date for the filter
4. No results will be displayed
5. Check the logs and you will see the query is using LIKE to search for the date

#### Steps to test this PR:
1. Apply this PR
2. Repeat the steps to reproduce the bug
3. This time the report will show data (if there is any)
4. If you check the logs you will see that the query is using the correct operator.

### CODE REVIEW

File: app/bundles/ReportBundle/Form/Type/DynamicFiltersType.php [hunk 1]
```diff
@@ -57,11 +57,18 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                        $args['widget'] = 'single_text';
+                        $args['format'] = 'y-M-d';
+                        $args['attr']['class'] .= ' datepicker';
+                        break;
                     case 'datetime':
                         $type           = 'datetime';
                         $args['input']  = 'string';
                         $args['widget'] = 'single_text';
-                        $args['format'] = 'Y-m-d H:i:s';
+                        $args['format'] = 'y-M-d H:i:s';
```
Comment (escopecz) [inline app/bundles/ReportBundle/Form/Type/DynamicFiltersType.php]: Are you sure about this?

`y | A two digit representation of a year`
`M | A short textual representation of a month, three letters`

So an example you want is `17-Jan-25 01:01:01`?

**Comment by escopecz** [comment]
There is one more issue I noticed. Set the dynamic date to some date and time. For example `2017-09-14 11:00:53` (maybe the seconds shouldn't be there or should be `00`). Then refresh the page and the dynamic filter date is `2017-9-14 11:` which is invalid format.

File: app/bundles/ReportBundle/Controller/ReportController.php [hunk 2]
```diff
@@ -630,8 +630,20 @@ public function viewAction($objectId, $reportPage = 1)
...
-        $filterSettings    = [];
+        /** @var array $dynamicFilters */
+        $dynamicFilters = $session->get('mautic.report.'.$objectId.'.filters', []);
+        $filterSettings = [];
+
+        if (count($dynamicFilters) > 0 && count($entity->getFilters()) > 0) {
+            foreach ($entity->getFilters() as $fid => $filter) {
+                foreach ($dynamicFilters as $dfcol => &$dfval) {
+                    if (1 === $filter['dynamic'] && $filter['column'] === $dfcol) {
+                        $dfval['expr'] = $filter['condition'];
```
Comment (Maxell92) [inline app/bundles/ReportBundle/Controller/ReportController.php]: 1) Are you sure you want to use same condition from filter?
2) Maybe add "break" to if condition?

File: app/bundles/ReportBundle/Controller/ReportController.php [hunk 3]
```diff
@@ -630,8 +630,20 @@ public function viewAction($objectId, $reportPage = 1)
...
         $filterDefinitions = $model->getFilterList($entity->getSource());
-        $dynamicFilters    = $session->get('mautic.report.'.$objectId.'.filters', []);
-        $filterSettings    = [];
+        /** @var array $dynamicFilters */
+        $dynamicFilters = $session->get('mautic.report.'.$objectId.'.filters', []);
+        $filterSettings = [];
+
+        if (count($dynamicFilters) > 0 && count($entity->getFilters()) > 0) {
+            foreach ($entity->getFilters() as $fid => $filter) {
+                foreach ($dynamicFilters as $dfcol => &$dfval) {
```
Comment (Maxell92) [inline app/bundles/ReportBundle/Controller/ReportController.php]: Do you have reason for using &? I would set it through an array, it is more readable for me.

**Comment by dongilbert** [comment]
@virlatinus Thoughts on the comments?

**Comment by virlatinus** [comment]
@escopecz The filter value issue was fixed

**Comment by javjim-mautic** [comment]
Working properly Tested.

**Pull request merged**